File: IAMenuController/IAMenuController.m
//
//  IAMenuController.m
//  IAMenuController
//
//  Created by Mark Adams on 12/16/11.
//  MIT License
//

#import <QuartzCore/QuartzCore.h>
#import "IAMenuController.h"
#import "IAScreenEdgeGestureRecognizer.h"

NSString *const IAMenuWillOpenNotification = @"IAMenuWillOpenNotification";
NSString *const IAMenuDidOpenNotification = @"IAMenuDidOpenNotification";
NSString *const IAMenuWillCloseNotification = @"IAMenuWillCloseNotification";
NSString *const IAMenuDidCloseNotification = @"IAMenuDidCloseNotification";

@interface IAMenuController ()

@property (nonatomic, strong) UIView *contentView;
@property (nonatomic, strong) UIView *tapInterceptView;

@end

#pragma mark -

@implementation IAMenuController

#pragma mark - Initialization

- (id)initWithMenuViewController:(UIViewController *)menu contentViewController:(UIViewController *)content;
{
    NSParameterAssert(menu && content);

    self = [super init];

    if (!self)
        return nil;

    _menuViewController = menu;
    _contentViewController = content;

    return self;
}

#pragma mark - Setters

- (void)setContentViewController:(UIViewController *)controller
{
    if (controller == _contentViewController)
        return;

    UIViewController *oldContent = _contentViewController;
    _contentViewController = controller;

    [self removeTapInterceptView];
    [oldContent willMoveToParentViewController:nil];

    [UIView animateWithDuration:0.2 delay:0.0 options:0 animations:^{
        self.contentView.frame = [self contentViewFrameForStaging];
    } completion:^(BOOL finished) {
        [oldContent.view removeFromSuperview];
        [oldContent removeFromParentViewController];

        [self addChildViewController:_contentViewController];
        [self.contentView addSubview:_contentViewController.view];
        [self resizeViewForContentView:_contentViewController.view];
        [_contentViewController didMoveToParentViewController:self];

        [UIView animateWithDuration:0.22 delay:0.1 options:0 animations:^{
            self.contentView.frame = [self contentViewFrameForClosedMenu];
        } completion:^(BOOL finished) {
            self.menuIsVisible = NO;
        }];
    }];
}

#pragma mark - UIViewController

- (NSUInteger)supportedInterfaceOrientations
{
    return UIInterfaceOrientationMaskPortrait;
}

#pragma mark - View Lifecycle

- (void)viewDidLoad
{
    [super viewDidLoad];
    [self setupViewControllers];
}

#pragma mark - Child Controller Setup
- (void)setupViewControllers
{
    [self setupContentViewController];
    [self setupMenuViewController];
}

- (void)setupContentViewController
{
    [self addChildViewController:self.contentViewController];
    [self.contentViewController viewWillAppear:NO];
    [self setupContentView];
    [self.contentViewController viewDidAppear:NO];
    [self.contentViewController didMoveToParentViewController:self];
}

- (void)setupContentView
{
    self.contentView = [[UIView alloc] initWithFrame:self.view.bounds];

    [self addPanGestureRecognizer];
    [self.contentView addSubview:self.contentViewController.view];
    [self resizeViewForContentView:self.contentViewController.view];
    [self.view addSubview:self.contentView];

    self.contentView.layer.shadowColor = [UIColor blackColor].CGColor;
    self.contentView.layer.shadowOffset = CGSizeMake(-2.0f, 0.0f);
    self.contentView.layer.shadowOpacity = 0.75f;
    self.contentView.layer.shadowPath = [UIBezierPath bezierPathWithRect:self.contentView.frame].CGPath;
    self.contentView.layer.shadowRadius = 3.0f;
}

- (void)setupMenuViewController
{
    [self addChildViewController:self.menuViewController];
    self.menuViewController.view.frame = self.view.bounds;
    [self.view insertSubview:self.menuViewController.view belowSubview:self.contentView];
    [self.menuViewController didMoveToParentViewController:self];
}

#pragma mark - Gesture Management

- (void)addPanGestureRecognizer
{
    IAScreenEdgeGestureRecognizer *pan = [[IAScreenEdgeGestureRecognizer alloc] initWithTarget:self action:@selector(pan:)];
    [self.contentView addGestureRecognizer:pan];
}

- (void)addTapInterceptView
{
    [self.contentView addSubview:self.tapInterceptView];
}

- (void)removeTapInterceptView
{
    [self.tapInterceptView removeFromSuperview];
}

- (UIView *)tapInterceptView {
    if (!_tapInterceptView) {
        CGRect frame = {{0, 0}, self.contentView.frame.size};
        _tapInterceptView = [[UIView alloc] initWithFrame:frame];
        _tapInterceptView.userInteractionEnabled = YES;
        UITapGestureRecognizer *tap = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(hideMenu)];
        [_tapInterceptView addGestureRecognizer:tap];
    }

    return _tapInterceptView;
}

#pragma mark - Menu State

- (void)menuWillOpen
{
    self.menuIsVisible = YES;
    [self.menuViewController viewWillAppear:YES];
    [[NSNotificationCenter defaultCenter] postNotificationName:IAMenuWillOpenNotification object:nil];
}

- (void)menuDidOpen
{
    self.menuIsVisible = YES;
    [self.menuViewController viewDidAppear:YES];
    [self addTapInterceptView];
    [[NSNotificationCenter defaultCenter] postNotificationName:IAMenuDidOpenNotification object:nil];
}

- (void)menuWillClose
{
    self.menuIsVisible = NO;
    [self.menuViewController viewWillDisappear:YES];
    [[NSNotificationCenter defaultCenter] postNotificationName:IAMenuWillCloseNotification object:nil];
}

- (void)menuDidClose
{
    self.menuIsVisible = NO;
    [self.menuViewController viewDidDisappear:YES];
    [self removeTapInterceptView];
    [[NSNotificationCenter defaultCenter] postNotificationName:IAMenuDidCloseNotification object:nil];
}

#pragma mark - Menu Presentation

- (void)toggleMenu
{
    (self.menuIsVisible)? [self hideMenu] : [self showMenu];
}

- (void)showMenu
{
    [self menuWillOpen];

    [[UIApplication sharedApplication] beginIgnoringInteractionEvents];

    [UIView animateWithDuration:0.225 delay:0 options:UIViewAnimationOptionCurveEaseInOut animations:^{
        self.contentView.frame = [self contentViewFrameForOpenMenu];
    } completion:^(BOOL finished) {
        [self menuDidOpen];
        [[UIApplication sharedApplication] endIgnoringInteractionEvents];
    }];
}

- (void)hideMenu
{
    [self menuWillClose];

    [[UIApplication sharedApplication] beginIgnoringInteractionEvents];

    [UIView animateWithDuration:0.225 delay:0 options:UIViewAnimationOptionCurveEaseInOut animations:^{
        self.contentView.frame = [self contentViewFrameForClosedMenu];
    } completion:^(BOOL finished) {
        [self menuDidClose];
        [[UIApplication sharedApplication] endIgnoringInteractionEvents];
    }];
}

#pragma mark - Pan Gesture Support
- (void)pan:(UIPanGestureRecognizer *)pan
{
    CGPoint translation = [pan translationInView:self.contentView];
    CGPoint velocity = [pan velocityInView:self.contentView];

    CGFloat minimumX = 0.0f;
    CGFloat maximumX = 276.0f;

    if (pan.state == UIGestureRecognizerStateBegan)
    {
        if (translation.x < minimumX)
            return;
    }
    else if (pan.state == UIGestureRecognizerStateChanged)
    {
        CGRect currentFrame = self.contentView.frame;
        CGFloat newX = currentFrame.origin.x + translation.x;

        if (newX < minimumX || newX > maximumX)
            return;

        currentFrame.origin.x = currentFrame.origin.x + translation.x;
        self.contentView.frame = currentFrame;
    }
    else if (pan.state == UIGestureRecognizerStateEnded)
    {
        if (CGRectGetMinX(self.contentView.frame) == 0.0f)
            return;

        CGFloat finalX = self.contentView.frame.origin.x + (0.55 * velocity.x);
        CGFloat travelDistance;

        if (finalX >= CGRectGetMidX(self.view.frame))
        {
            finalX = maximumX;
            travelDistance = maximumX - CGRectGetMinX(self.contentView.frame);
            [self menuWillOpen];
        }
        else
        {
            finalX = minimumX;
            travelDistance = minimumX + CGRectGetMinX(self.contentView.frame);
            [self menuWillClose];
        }

        NSTimeInterval duration = (travelDistance / ABS(velocity.x));

        if (duration < 0.1) {
            duration = 0.1;
        } else if (duration > 0.25) {
            duration = 0.25;
        }

        [UIView animateWithDuration:duration delay:0.0 options:UIViewAnimationOptionCurveEaseOut animations:^{
            CGRect frame = self.contentView.frame;
            frame.origin.x = finalX;
            self.contentView.frame = frame;
        } completion:^(BOOL finished) {
            if (finalX == maximumX)
            {
                [self menuDidOpen];
            }
            else if (finalX == 0.0f)
            {
                [self menuDidClose];
            }
        }];
    }

    [pan setTranslation:CGPointZero inView:self.contentView];
}

#pragma mark - Frame Calculation
- (CGRect)contentViewFrameForOpenMenu
{
    CGRect frame = self.contentView.frame;
    frame.origin.x = 276.0f;

    return frame;
}

- (CGRect)contentViewFrameForClosedMenu
{
    CGRect frame = self.contentView.frame;
    frame.origin.x = 0.0f;

    return frame;
}

- (CGRect)contentViewFrameForStaging
{
    CGRect frame = self.contentView.frame;
    frame.origin.x = CGRectGetMaxX(self.view.frame) + 44.0f;

    return frame;
}

- (void)resizeViewForContentView:(UIView *)view
{
    view.frame = self.view.bounds;
}

@end

@implementation UIViewController (IAMenuController)

- (IAMenuController *)menuController
{
    if (!self.parentViewController) return nil;

    if ([self.parentViewController isKindOfClass:[IAMenuController class]])
        return (IAMenuController *)self.parentViewController;

    return [self.parentViewController menuController];
}

@end

File: IAMenuController/IAMenuController.swift
/*  IAMenuController.swift
 *  IAMenuController
 *
 *  Converted to Swift by Taylor Wright-Sanson on 9/7/18.
 *  Created by Mark Adams on 12/16/11.
 *  MIT License
 */

import UIKit
import QuartzCore

@objc (IAMenuController)
open class IAMenuController: UIViewController {

  // MARK: -

  @objc open let menuViewController: UIViewController
  @objc open var contentViewController: UIViewController {
    willSet {
      let oldContent = self.contentViewController

      removeTapInterceptView()
      oldContent.willMove(toParentViewController: nil)

      UIView.animate(withDuration: 0.2,
                     animations: {
                      self.contentView.frame = self.contentViewFrameForStaging()
      }) { (finished) in
        oldContent.view.removeFromSuperview()
        oldContent.removeFromParentViewController()

        self.addChildViewController(newValue)
        self.contentView.addSubview(newValue.view)
        self.resizeViewFor(contentView: newValue.view)
        newValue.didMove(toParentViewController: self)

        UIView.animate(withDuration: 0.22,
                       delay: 0.1,
                       options: UIViewAnimationOptions.curveLinear,
                       animations: {
                        self.contentView.frame = self.contentViewFrameForClosedMenu()
        }, completion: { (finished) in
          self.menuIsVisible = false
        })
      }
    }
  }

  @objc open var contentView: UIView
  @objc open var menuIsVisible: Bool = false
  @objc open let percentageOfScreenWidthUsedByMenu: CGFloat = 86.25
  private var tapInterceptView: UIView?

  // MARK: - Initialization

  required public init?(coder aDecoder: NSCoder) {
    fatalError("init(coder:) has not been implemented")
  }

  @objc public init(menuViewController: UIViewController, contentViewController: UIViewController) {
    self.menuViewController = menuViewController
    self.contentViewController = contentViewController
    self.contentView = UIView(frame: .zero)

    super.init(nibName: nil, bundle: nil)
  }

  // MARK: - UIViewController

  override open var supportedInterfaceOrientations: UIInterfaceOrientationMask {
    return .portrait
  }

  // MARK: - View Lifecycle

  override open func viewDidLoad() {
    super.viewDidLoad()
    setupViewControllers()
  }

  // MARK: - Child Controller Setup

  private func setupContentView() {
    contentView = UIView(frame: view.bounds)

    addPanGestureRecognizer()
    contentView.addSubview(contentViewController.view)
    resizeViewFor(contentView: contentViewController.view)
    view.addSubview(contentView)

    contentView.layer.shadowColor = UIColor.black.cgColor
    contentView.layer.shadowOffset = CGSize(width: -2.0, height: 0.0)
    contentView.layer.shadowOpacity = 0.75
    contentView.layer.shadowPath = UIBezierPath(rect: contentView.frame).cgPath
    contentView.layer.shadowRadius = 3.0
  }

  private func setupContentViewController() {
    addChildViewController(contentViewController)
    contentViewController.viewWillAppear(false)
    setupContentView()
    contentViewController.viewDidAppear(false)
    contentViewController.didMove(toParentViewController: self)
  }

  private func setupMenuViewController() {
    addChildViewController(menuViewController)
    menuViewController.view.frame = view.bounds
    view.insertSubview(menuViewController.view, belowSubview: contentView)
    menuViewController.didMove(toParentViewController: self)
  }

  private func setupTapInterceptView() {
    if tapInterceptView == nil {
      tapInterceptView = UIView(frame: CGRect(origin: CGPoint(x: 0, y: 0), size: contentView.frame.size))
      tapInterceptView!.isUserInteractionEnabled = true
      let tap = UITapGestureRecognizer(target: self, action: #selector(hideMenu))
      tapInterceptView!.addGestureRecognizer(tap)
    }
  }

  private func setupViewControllers() {
    setupContentViewController()
    setupMenuViewController()
  }

  // MARK: - Gesture Management

  private func addPanGestureRecognizer() {
    let pan: IAScreenEdgeGestureRecognizer = IAScreenEdgeGestureRecognizer(target: self, action: #selector(pan(_:)))
    contentView.addGestureRecognizer(pan)
  }

  private func addTapInterceptView() {
    if tapInterceptView == nil {
      setupTapInterceptView()
    }

    contentView.addSubview(tapInterceptView!)
  }

  private func removeTapInterceptView() {
    if tapInterceptView == nil {
      setupTapInterceptView()
    }

    tapInterceptView!.removeFromSuperview()
  }

  // MARK: - Menu State

  private func menuDidClose() {
    menuIsVisible = false
    menuViewController.viewDidDisappear(true)
    removeTapInterceptView()
    NotificationCenter.default.post(name: .IAMenuDidClose, object: nil)
  }

  private func menuDidOpen() {
    menuIsVisible = true
    menuViewController.viewDidAppear(true)
    addTapInterceptView()
    NotificationCenter.default.post(name: .IAMenuDidOpen, object: nil)
  }

  private func menuWillClose() {
    menuIsVisible = false
    menuViewController.viewWillDisappear(true)
    NotificationCenter.default.post(name: .IAMenuWillClose, object: nil)
  }

  private func menuWillOpen() {
    menuIsVisible = true
    menuViewController.viewWillAppear(true)
    NotificationCenter.default.post(name: .IAMenuWillOpen, object: nil)
  }

  // MARK: - Menu Presentation - Open

  @objc open func toggleMenu() {
    menuIsVisible ? hideMenu() : showMenu()
  }

  // MARK: - Menu Presentation - Private

  @objc private func hideMenu() {
    menuWillClose()
    UIApplication.shared.beginIgnoringInteractionEvents()

    UIView.animate(withDuration: 0.225,
                   delay: 0,
                   options: .curveEaseInOut,
                   animations: {
                    self.contentView.frame = self.contentViewFrameForClosedMenu()
    }) { (finished) in
      self.menuDidClose()
      UIApplication.shared.endIgnoringInteractionEvents()
    }
  }

  private func showMenu() {
    menuWillOpen()
    UIApplication.shared.beginIgnoringInteractionEvents()

    UIView.animate(withDuration: 0.225,
                   delay: 0,
                   options: .curveEaseInOut,
                   animations: {
                    self.contentView.frame = self.contentViewFrameForOpenMenu()
    }) { (finished) in
      self.menuDidOpen()
      UIApplication.shared.endIgnoringInteractionEvents()
    }
  }

  // MARK: - Pan Gesture Support

  @objc private func pan(_ pan: UIPanGestureRecognizer) {
    let translation = pan.translation(in: contentView)
    let velocity = pan.velocity(in: contentView)

    let minimumX: CGFloat = 0.0
    let maximumX: CGFloat = menuWidth()

    if pan.state == .began {
      return
    } else if pan.state == .changed {
      var currentFrame = contentView.frame
      let newX = currentFrame.origin.x + translation.x

      if newX < minimumX, newX > maximumX { return }

      currentFrame.origin.x = currentFrame.origin.x + translation.x
      contentView.frame = currentFrame
    } else if pan.state == .ended {
      if contentView.frame.minX == 0.0 { return }

      var finalX = contentView.frame.origin.x + (0.55 * velocity.y)
      var travelDistance: CGFloat = 0.0

      if finalX >= view.frame.midX {
        finalX = maximumX
        travelDistance = maximumX - contentView.frame.minX
        menuWillOpen()
      } else {
        finalX = minimumX
        travelDistance = minimumX + contentView.frame.minX
        menuWillClose()
      }

      var duration = travelDistance / abs(velocity.x)

      if duration < 0.1 {
        duration = 0.1
      } else if duration > 0.25 {
        duration = 0.25
      }

      UIView.animate(withDuration: TimeInterval(duration),
                     delay: 0,
                     options: .curveEaseOut,
                     animations: {
                      var frame = self.contentView.frame
                      frame.origin.x = finalX
                      self.contentView.frame = frame
      }) { (finished) in
        if finalX == maximumX {
          self.menuDidOpen()
        } else if finalX == 0.0 {
          self.menuDidClose()
        }
      }
    }

    pan.setTranslation(.zero, in: self.contentView)
  }

  // MARK: - Frame Calculation

  private func contentViewFrameForClosedMenu() -> CGRect {
    var frame = self.contentView.frame
    frame.origin.x = 0.0

    return frame
  }

  private func contentViewFrameForStaging() -> CGRect {
    var frame = self.contentView.frame
    frame.origin.x = view.frame.maxX + UIScreen.main.bounds.size.width - menuWidth()

    return frame
  }

  private func contentViewFrameForOpenMenu() -> CGRect {
    var frame = self.contentView.frame
    frame.origin.x = menuWidth()

    return frame
  }

  private func menuWidth() -> CGFloat {
    return UIScreen.main.bounds.size.width * percentageOfScreenWidthUsedByMenu / 100.0
  }

  private func resizeViewFor(contentView: UIView) {
    contentView.frame = view.bounds
  }

  // MARK: - Open Methods For Objc Availability

  @objc open class func iAMenuWillOpenNotification() -> String { return IAMenuWillOpenNotification }
  @objc open class func iAMenuDidOpenNotification() -> String { return IAMenuDidOpenNotification }
  @objc open class func iAMenuWillCloseNotification() -> String { return IAMenuWillCloseNotification }
  @objc open class func iAMenuDidCloseNotification() -> String { return IAMenuDidCloseNotification }
}

// MARK: - Notification Constants

private let IAMenuWillOpenNotification: String = "IAMenuWillOpenNotification"
private let IAMenuDidOpenNotification: String = "IAMenuDidOpenNotification"
private let IAMenuWillCloseNotification: String = "IAMenuWillCloseNotification"
private let IAMenuDidCloseNotification: String = "IAMenuDidCloseNotification"

public extension Notification.Name {
  static let IAMenuWillOpen = Notification.Name(IAMenuWillOpenNotification)
  static let IAMenuDidOpen = Notification.Name(IAMenuDidOpenNotification)
  static let IAMenuWillClose = Notification.Name(IAMenuWillCloseNotification)
  static let IAMenuDidClose = Notification.Name(IAMenuDidCloseNotification)
}

// MARK: - UIViewController Extension

extension UIViewController {
  @objc open func menuController() -> IAMenuController? {
    guard let parent = parent else { return nil }

    if parent.isKind(of: IAMenuController.self) {
      return parent as? IAMenuController
    }

    return parent.menuController()
  }
}
